File: Sort/merge_sort.py
def merge_sort(arr):
  
    if len(arr)> 1:
        mid = len(arr)//2
        arr1 = arr[:mid]
        arr2 = arr[mid:]
        merge_sort(arr1)
        merge_sort(arr2)

        i = j = k = 0
        while i < len(arr1) and j < len(arr2):
            if arr1[i] < arr2[j]:
                arr[k] = arr1[i]
                i += 1
            else:
                arr[k] = arr2[j]
                j += 1
            k += 1

        while i < len(arr1):
            arr[k] = arr1[i]
            i += 1
            k += 1
        while j < len(arr2):
            arr[k] = arr2[j]
            j += 1
            k += 1

    return arr


if __name__ == '__main__':
    array = [5, 1, 7, 3, 6, 1, 3, 9]
    ls = merge_sort(array)
    print(ls)

File: algorithm/merge_sort.cpp
#include <iostream>
using namespace std;

void merge(int arr[], int f, int m, int l)
{
    int s1 = m - f + 1;
    int s2 = l - m;
    int a1[s1], a2[s2];

    for (int i = 0; i < s1; i++)
        a1[i] = arr[f + i];
    for (int j = 0; j < s2; j++)
        a2[j] = arr[m + j + 1];

    int i, j, k;
    i = 0, j = 0;
    k = f;
    while (i < s1 && j < s2)
    {
        if (a1[i] <= a2[j])
        {
            arr[k] = a1[i];
            i++;
        }
        else
        {
            arr[k] = a2[j];
            j++;
        }
        k++;
    }

    while (i < s1)
    {
        arr[k] = a1[i];
        i++;
        k++;
    }
    while (j < s2)
    {
        arr[k] = a2[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int f, int l)
{
    int m = (f + l) / 2;
    if (f < l)
    {
        mergeSort(arr, f, m);
        mergeSort(arr, m + 1, l);
        merge(arr, f, m, l);
    }
}

int main()
{
    int arr[] = {1, 3, 8, 9, 4, 3, 2, 1};
    int n = sizeof(arr) / sizeof(arr[0]);
    mergeSort(arr, 0, n - 1);

    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
}